### ISSUE TITLE
[Users help needed] Backups of PineBuds Pro's pre-installed firmware for research

### ISSUE THREAD

**Issue by lizz-34** [pine64/OpenPineBuds#31]
The developers are trying to get ANC work, and it would be useful to get backups of the closed firmware pre-installed in PineBuds Pro with a large sampling. Of course, the pre-installed firmware binary is shared [here](https://wiki.pine64.org/wiki/PineBuds_Pro#Firmware_Images_Build), but Pine Store calibrates the ANC individually for each device. Developers are interested of this calibration.

Share the backup of pre-installed firmware in your both buds here in @Ralim's Google Drive: https://drive.google.com/drive/folders/1Ufo-ovEmK3cEiX5hJfNgJgkyBKkfru-F

**Comment by WebWire-NL** [comment]
Are you sure about the statement that every earbud is calibrated per device?
As far as i know the device is calibrated on the hardware and the original software is just calibrated on the hardware of the earbuds, we only need to find the right calibration values and binary's for the specific hardware which is used in the original firmware. Seems there's a specific piece of proprietary firmware we are missing to control the ANC chip.

**Comment by Ralim** [comment]
Every single earbud Pine makes is individually run through the calibration machine and the values are stored into the device.

While the equipment can do a lot of measurements we don't know exactly what it is calibrating.

The settings area definitely changes more than just the MAC address across buds. So the idea here is to figure out the flash format and parse the diffs to find out how much of a change the calibration is.

We hope that either it's near enough that per device calibration isn't required. Or if it is different we will need to make tooling to convert it for the open source firmware.

**Comment by gamelaster** [comment]
Also note, as far as I know, there is also automatic calibration option, which will factory firmware do, if "external factory calibration" is not preset, and it should have good results.

**Comment by lizz-34** [comment]
@WebWire-NL

> Are you sure about the statement that every earbud is calibrated per device?

Tbh no info, but Ralim and others told me about the calibration in the Matrix room and I believe them.

> the device is calibrated on the hardware and the original software is just calibrated on the hardware of the earbuds

?

> we only need to find the right calibration values and binary's for the specific hardware which is used in the original firmware

?

>  Seems there's a specific piece of proprietary firmware we are missing to control the ANC chip.

What piece missing from the backup?

---

@Ralim

> While the equipment can do a lot of measurements we don't know exactly what it is calibrating.

Are you still talking about the calibration machine at the Pine factory?

> The settings area definitely changes more than just the MAC address across buds. So the idea here is to figure out the flash format and parse the diffs to find out how much of a change the calibration is.

Is MAC address stored into Flash memory? Isn't it stored into ROM?

> We hope that either it's near enough that per device calibration isn't required.

You assume that the average of all calibrations is enough good?

---

@gamelaster 

> Also note, as far as I know, there is also automatic calibration option

Like [this](https://youtu.be/ILhD50uLJMs)?

>  which will factory firmware do, if "external factory calibration" is not preset, and it should have good results.

Source for this?

**Comment by gamelaster** [comment]
@lizz-34 no idea how it works, but I think it is fully automatical process. Source for all this information are info provided to me by Pine Store folks. I'm 100% sure that every bud is separately calibrated at factory.

**Comment by JF002** [comment]
I've just added my contribution to the drive :)

**Comment by gamelaster** [comment]
@ThatcherC it is very possible that @Ralim have first batch device.

**Comment by WebWire-NL** [comment]
Anyone has any instructions to backup on windows?
Got 2 sets pre-ordered before it was even available.

**Comment by ThatcherC** [comment]
Using `binwalk` to run a diff on the binaries, my image and @JF002's only differ on 40 or so lines (16 bytes per line) of the binwalk output, which is pretty small considing the 4MB image size. Here's the command I used to get the binwalk diff:

```
$ binwalk -Wi user-fws/thatcherc-firmware-0-1672681053.bin.bkp user-fws/thatcherc-firmware-1-1672681053.bin.bkp
```

**Comment by WebWire-NL** [comment]
@ThatcherC Also uploaded mine, two different pairs of buds. Spent a lot of time in WSL2 :) sos didn't really try to analyze them. Seems they're November 2022 build also.

**Comment by Ralim** [comment]
Yeah my buds is an earlier batch so not surprised its different tbqh.
Was hoping otherse had a matching batch to make it easier to figure out :)

**Comment by JF002** [comment]
I noticed that each time I run the backup script, the backup files are a bit different (a few lines are different at the end of the file). Is this expected?

**Comment by ThatcherC** [comment]
@JF002 I noticed the same thing - a handful of bytes are different at the end of the file. It came up on the Discord too and the conclusion is that it's probably some state or connection information that the Buds store, like Bluetooth info about the phone/computer they're connected to.

**Comment by zjstraus** [comment]
I just added another dump from a pair with Nov 2nd build dates; taken straight out of the package.

**Comment by futurl** [comment]
I would love to contribute a firmware backup.
How do I upload it to Ralims Google Drive (user rights needed)?

**Comment by FintasticMan** [comment]
Just added mine myself, I didn't need any permissions, just a Google account.

**Comment by futurl** [comment]
I do not have google. Anybody willing to upload my files?
[futurl firmware-1677959367-.zip](https://github.com/pine64/OpenPineBuds/files/10889609/futurl.firmware-1677959367-.zip)

**Comment by FintasticMan** [comment]
No problem! Done.

**Comment by jdc-cunningham** [comment]
Added (jcun4128)

Got my Pinebuds Pro on Apr 10, 2023

**Comment by daygr** [comment]
Added, purchased April 28, 2023

**Comment by nik012003** [comment]
Added, purchased May 24.  2023

**Comment by SethFalco** [comment]
I've included my backup, purchased on 09 January 2023.

**Comment by shymega** [comment]
I'm wondering if it's best, to avoid cluttering notifications, for when people upload their device info to the GDrive, they react to this comment with `:+1:`, and this way we don't get bombarded with notifications.

This should have been done from the beginning, to be honest, but it's something I've noticed.

**Comment by Yotson** [comment]
I also don't have a Google account. Here are mine.
[firmware-1689501067.zip](https://github.com/pine64/OpenPineBuds/files/12063743/firmware-1689501067.zip)

**Comment by hamblingreen1** [comment]
Here are the firmware backups for PineBuds that I purchased on November 9, 2023. I purchased only the buds as replacement parts, as I already had the case. I'm not sure whether that'll be relevant, but I thought I'd include it anyway.
[firmware-1700097469.zip](https://github.com/pine64/OpenPineBuds/files/13383017/firmware-1700097469.zip)

**Comment by nrbuster** [comment]
Attaching my buds default firmware.
Received these on 2/18/2024, only just now changing my firmware out.
[firmware-backups.tar.gz](https://github.com/user-attachments/files/15976399/firmware-backups.tar.gz)

**Comment by austingarrigus** [comment]
Here is my factory firmware.
[firmware-backups.tar.gz](https://github.com/user-attachments/files/16289103/firmware-backups.tar.gz)
